File: Data Structures & Algorithms/daily-temperatures/submission-0.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int[] result = new int[temperatures.length];
        Stack<int[]> stack = new Stack<>(); // [temp, index]
        for (int i = 0; i < temperatures.length; i++) {
            while (!stack.isEmpty() && temperatures[i] > stack.peek()[0]) {
                int[] curr = stack.pop();
                result[curr[1]] = i - curr[1];
            }
            stack.push(new int[]{temperatures[i], i});
        }
        return result;
    }
}

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = []
        result = 0

        for i in range(len(heights)):
            start = i

            while stack and stack[-1][1] > heights[i]:
                index, height = stack.pop()
                result = max(result, (i - index) * height)
                start = index

            stack.append((start, heights[i]))

        while stack:
            index, height = stack.pop()
            result = max(result, (len(heights) - index) * height)

        return result